File: python/numbers/numbers.py
import numpy as np

def close(a:np.number, b:np.number, rtol=1e-05, atol=1e-08):
    """
    Checks whether two given numbers are close to each other within the given tolerances
    :param a: First number
    :param b: Second number, used as reference
    :param rtol: Relative tolerance, which depends on the value b
    :param atol: Absolute tolerance
    :return: Whether the numbers are close to each other
    """
    return abs(a - b) <= atol + rtol * abs(b)

File: c/numbers/numbers.c
#include "numbers.h"

#include <math.h>

double numbers_compute_epsilon() {
    double eps = 1.0;
    const double one = 1.0;

    while (one + eps != one) {
        eps = eps / 2.0;
    }

    eps *= 2;

    return eps;
}

bool numbers_is_close(const double a, const double b, const double rtol, const double atol) {
    return fabs(a - b) <= atol + rtol * fabs(b);
}